File: src/Oikono.Web/oikono/src/pages/recipes/lib/recipe.ts
export type IngredientRef = {
    id: string
    name: string
}
export type PartIngredient = {
    ingredient: IngredientRef
    amount: number
    unit: string
}

export type MergedIngredient = {
    name: string
    unit: string
    amount: number
}

/**
 * Merge a list of ingredients by summing amounts for identical name+unit pairs.
 * - Case-insensitive, trims whitespace
 * - Only sums if unit string matches (after trim, case-insensitive)
 */
export function mergeAndSumIngredients(ings?: PartIngredient[] | null): MergedIngredient[] {
    const map = new Map<string, MergedIngredient>();
    for (const ing of ings ?? []) {
        const name = (ing.ingredient?.name ?? "").trim();
        const unit = (ing.unit ?? "").trim();
        const key = `${name.toLowerCase()}__${unit.toLowerCase()}`;
        const prev = map.get(key);
        if (prev) {
            prev.amount += ing.amount;
        } else {
            map.set(key, { name, unit, amount: ing.amount });
        }
    }
    return Array.from(map.values());
}

export type Device = {
    id: string
    name: string
}

export type RecipeStep = {
    description: string
    duration?: number         // Minuten
    device?: Device
    temperature?: number      // °C, falls relevant
}

export type IngredientPart = {
    name: string | null
    ingredients: PartIngredient[]
}

export type Asset = {
    id: string,
    fileName: string,
    contentType: string,
    data: string, // base64
}

export type SideDish = {
    id: string
    name: string
}

export type Tag = {
    id: string
    name: string
}

export type Recipe = {
    id: string
    name: string
    images: Asset[]          // mehrere Bilder
    parts: IngredientPart[]  // optionale Gruppen von Zutaten (z. B. Teig, Füllung, Crumble)
    steps: RecipeStep[]
    tags: Tag[]            // z. B. ["vegan", "schnell"]

    portions: number          // Pflichtfeld
    calories?: number         // optional, kcal pro Portion
    sideDishes?: SideDish[]     // optional, z. B. ["Reis", "Salat"]
    rating?: number           // ganzzahlig 0-5
}

File: src/Oikono.Domain/Recipes/Recipe.cs
using Oikono.Domain.Assets;
using Oikono.Domain.Models;
using Oikono.Domain.Recipes.ValueObjects;
using Oikono.Domain.Users;
using Oikono.Domain.Users.ValueObjects;

namespace Oikono.Domain.Recipes;

public class Recipe : AggregateRoot<RecipeId>
{
    private readonly List<Asset> _images = [];
    private readonly List<Part> _parts = [];
    private readonly List<Step> _steps = [];
    private readonly List<Tag> _tags = [];
    private readonly List<SideDish> _sideDishes = [];

    public Recipe(string name, int portions,  int? calories, int? rating)
    {
        Name = name;
        Portions = portions;
        Calories = calories;
        Rating = rating;
    }

    public string Name { get; private set; }
    public int Portions { get; private set; }
    public int? Calories { get; private set; }
    public int? Rating { get; private set; }

    // Bilder
    public IReadOnlyList<Asset> Images => _images.AsReadOnly();

    // Teile (z. B. Teig, Füllung)
    public IReadOnlyList<Part> Parts => _parts.AsReadOnly();

    // Zubereitungsschritte
    public IReadOnlyList<Step> Steps => _steps.AsReadOnly();

    // Tags
    public IReadOnlyList<Tag> Tags => _tags.AsReadOnly();

    // Beilagen
    public IReadOnlyList<SideDish> SideDishes => _sideDishes.AsReadOnly();

    public void AddSideDish(SideDish sideDish)
    {
        if (!_sideDishes.Any(sd => sd.Id == sideDish.Id))
            _sideDishes.Add(sideDish);
    }

    public void AddPart(Part part)
    {
        _parts.Add(part);
    }

    public void AddStep(Step step)
    {
        _steps.Add(step);
    }

    public void AddImage(Asset image)
    {
        _images.Add(image);
    }

    public void AddTag(Tag tag)
    {
        _tags.Add(tag);
    }
}